torchao/quantization/quantize_/workflows/int4/int4_plain_int32_tensor.py

```diff
@@ -0,0 +1,181 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import List
+
+import torch
+
+from torchao.quantization.quant_primitives import (
+    MappingType,
+    _choose_qparams_affine,
+    _quantize_affine,
+)
+from torchao.utils import (
+    TorchAOBaseTensor,
+)
+
+__all__ = [
+    "Int4PlainInt32",
+]
+
+aten = torch.ops.aten
+
+
+class Int4PlainInt32(TorchAOBaseTensor):
+    """
+    int4 weight-only quantization on XPU with oneDNN as backend (groupwise quantization only)
+
+    Tensor Attributes:
+        qdata: packed int4 weigh, always viewed as a 2D (N, K/2) tensor
+        scale: (K/group_size, N), dtype is the same as the original Tensor dtype
+        zero_point: (K/group_size, N)
+
+    Non-Tensor Attributes:
+        block_size: the block size for quantization, representing the granularity, for groupwise quantization, will have block_size (1, group_size).
+                    we only support group_size = 32/64/128.
+        shape: shape of the original Tensor
+
+    """
+
+    tensor_data_names = ["qdata", "scale", "zero_point"]
+    tensor_attribute_names = ["block_size", "shape"]
+
+    def __new__(
+        cls,
+        qdata,
+        scale,
+        zero_point,
+        block_size,
+        shape,
+    ):
+        kwargs = {}
+        kwargs["device"] = qdata.device
+        kwargs["dtype"] = scale.dtype
+        kwargs["requires_grad"] = False
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(self, qdata, scale, zero_point, block_size, shape):
+        self.qdata = qdata
+        self.scale = scale
+        self.zero_point = zero_point
+        self.block_size = block_size
+
+    def _quantization_type(self):
+        return f"shape={self.shape}, block_size={self.block_size}, device={self.device}"
+
+    @classmethod
+    def from_hp(
+        cls,
+        w: torch.Tensor,
+        block_size: List[int],
+    ):
+        assert w.ndim == 2 and w.device.type == "xpu", (
+            f"Expecting 2D tensor on XPU, but got: {w.shape} on {w.device.type}"
+        )
+        assert len(block_size) == w.ndim
+
+        original_shape = w.shape
+        mapping_type = MappingType.ASYMMETRIC
+        target_dtype = torch.int32
+        quant_min = 0
+        quant_max = 15
+        eps = 1e-6
+        scale_dtype = None
+        zero_point_dtype = torch.int32
+        scale, zero_point = _choose_qparams_affine(
```

Comment:
nit: we can use the official op? https://github.com/pytorch/ao/blob/f02354d82463e21e760b1ea5b03ee448f3ee6fe3/torchao/quantization/quant_primitives.py#L1194 same for quantize_affine

Reply:
Done.